src/util/bip143.rs

```diff
@@ -46,27 +46,27 @@ impl SighashComponents {
     /// script_sig and witnesses.
     pub fn new(tx: &Transaction) -> SighashComponents {
         let hash_prevouts = {
-            let mut enc = SigHash::engine();
+            let mut enc = sha256d::Hash::engine();
             for txin in &tx.input {
                 txin.previous_output.consensus_encode(&mut enc).unwrap();
             }
-            SigHash::from_engine(enc)
+            sha256d::Hash::from_engine(enc)
         };
 
         let hash_sequence = {
-            let mut enc = SigHash::engine();
+            let mut enc = sha256d::Hash::engine();
             for txin in &tx.input {
                 txin.sequence.consensus_encode(&mut enc).unwrap();
             }
-            SigHash::from_engine(enc)
+            sha256d::Hash::from_engine(enc)
         };
 
         let hash_outputs = {
-            let mut enc = SigHash::engine();
+            let mut enc = sha256d::Hash::engine();
             for txout in &tx.output {
                 txout.consensus_encode(&mut enc).unwrap();
             }
-            SigHash::from_engine(enc)
+            sha256d::Hash::from_engine(enc)
```

Comment:
I got your argument why engines are better, however it really seems that we have a plenty of repeated code with the current approach. Moving all patterns of `enc=engine....;consensus_encode(engine);...from_engine(enc)` into a function in `Hash` trait seems to be reasonable. Probably this should be done in separate PR on which I can work today

Reply:
That's usually the case in things that use Read/Write APIs 
```
let mut f: File;
let mut buf = Vec::new();
f.read(&mut but)?;
```

Reply:
I will probably still suggest generic programming best practices (avoid repeated code) over Rust std lib API strange parts, which may be just still undersigned at this stage...

Reply:
See my comment here: https://github.com/rust-bitcoin/rust-bitcoin/pull/388#issuecomment-574125643